File: packages/cleat-as/scripts/inject-metadata.js
#!/usr/bin/env node
/**
 * inject-metadata.js — Post-compile metadata stamping for AS WASM workflows.
 *
 * Reads a compiled WASM binary and injects (or replaces) the "cleat.metadata"
 * custom section (section ID 0x00) with workflow identity information.
 *
 * Usage:
 *   node scripts/inject-metadata.js <wasm-file> \
 *     --name PlaceOrder \
 *     --version 3 \
 *     --min-version 1 \
 *     --abi-version 1 \
 *     --plugin-deps '{"llm":">=1.2.0"}' \
 *     [--output <output.wasm>]
 *
 * Environment variable alternative:
 *   CLEAT_WORKFLOW_NAME=PlaceOrder \
 *   CLEAT_WORKFLOW_VERSION=3 \
 *   CLEAT_MIN_COMPATIBLE_VERSION=1 \
 *   CLEAT_ABI_VERSION=1 \
 *   CLEAT_PLUGIN_DEPS='{"llm":">=1.2.0"}' \
 *   node scripts/inject-metadata.js <wasm-file>
 *
 * This is a post-compile step — run AFTER "asc" has produced the .wasm file.
 * It requires no external WASM libraries — pure Node.js Buffer manipulation.
 */

const fs = require("fs");
const path = require("path");

const SECTION_NAME = "cleat.metadata";
const WASM_MAGIC = Buffer.from([0x00, 0x61, 0x73, 0x6D]); // \0asm
const WASM_VERSION = Buffer.from([0x01, 0x00, 0x00, 0x00]); // v1

/**
 * Encode a number as unsigned LEB128 varint.
 */
function encodeULEB128(value) {
  const bytes = [];
  do {
    let b = value & 0x7f;
    value >>>= 7;
    if (value !== 0) b |= 0x80;
    bytes.push(b);
  } while (value !== 0);
  return Buffer.from(bytes);
}

/**
 * Decode an unsigned LEB128 varint at the given offset.
 * Returns { value, bytesRead }.
 */
function decodeULEB128(buf, offset) {
  let result = 0;
  let shift = 0;
  let bytesRead = 0;

  for (let i = offset; i < buf.length; i++) {
    bytesRead++;
    const b = buf[i];
    result |= (b & 0x7f) << shift;
    if ((b & 0x80) === 0) {
      return { value: result, bytesRead };
    }
    shift += 7;
    if (shift > 63) {
      throw new Error("invalid WASM binary: corrupted section length encoding");
    }
  }
  throw new Error("invalid WASM binary: truncated section length encoding");
}

/**
 * Build a complete custom section.
 * Section format: [section_id: 0x00] [payload_len: LEB128] [name_len: LEB128] [name] [payload]
 */
function buildCustomSection(name, payload) {
  const nameBytes = Buffer.from(name, "utf8");
  const nameLen = encodeULEB128(nameBytes.length);
  const fullPayload = Buffer.concat([nameLen, nameBytes, payload]);
  const payloadLen = encodeULEB128(fullPayload.length);
  return Buffer.concat([Buffer.from([0x00]), payloadLen, fullPayload]);
}

/**
 * Find a custom section in a WASM binary.
 * Returns { payload, sectionStart, sectionEnd } or null if not found.
 */
function findCustomSection(wasmBuf, name, filePath = "") {
  if (wasmBuf.length < 8) return null;

  // Validate magic.
  if (wasmBuf.slice(0, 4).compare(WASM_MAGIC) !== 0) {
    throw new Error("invalid WASM binary: bad magic number in '" + filePath + "'");
  }
  if (wasmBuf.slice(4, 8).compare(WASM_VERSION) !== 0) {
    throw new Error("invalid WASM binary: unsupported version '" + wasmBuf.slice(4, 8).toString("hex") + "' (only v1 supported) in '" + filePath + "'");
  }

  const nameBytes = Buffer.from(name, "utf8");
  let offset = 8;

  while (offset < wasmBuf.length) {
    const sectionStart = offset;
    const sectionId = wasmBuf[offset];
    offset++;

    if (offset >= wasmBuf.length) break;

    const { value: payloadLen, bytesRead } = decodeULEB128(wasmBuf, offset);
    offset += bytesRead;
    const sectionEnd = offset + payloadLen;

    if (offset + payloadLen > wasmBuf.length) {
      throw new Error("Section extends beyond end of WASM binary");
    }

    if (sectionId === 0) {
      // Custom section — check the name.
      const { value: nameLen, bytesRead: nameBytesRead } = decodeULEB128(wasmBuf, offset);
      if (nameBytesRead > 0) {
        const nameStart = offset + nameBytesRead;
        const nameEnd = nameStart + nameLen;
        if (nameEnd <= offset + payloadLen) {
          const sectionName = wasmBuf.slice(nameStart, nameEnd);
          if (sectionName.compare(nameBytes) === 0) {
            return {
              payload: wasmBuf.slice(nameEnd, sectionEnd),
              sectionStart,
              sectionEnd,
            };
          }
        }
      }
    }

    offset += payloadLen;
  }

  return null;
}

/**
 * Inject (or replace) the cleat.metadata custom section in a WASM binary.
 */
function injectMetadata(wasmBytes, meta, filePath = "") {
  const jsonPayload = Buffer.from(JSON.stringify(meta), "utf8");

  // Remove existing cleat.metadata section if present.
  const existing = findCustomSection(wasmBytes, SECTION_NAME, filePath);
  let base = wasmBytes;
  if (existing) {
    base = Buffer.concat([
      wasmBytes.slice(0, existing.sectionStart),
      wasmBytes.slice(existing.sectionEnd),
    ]);
  }

  const newSection = buildCustomSection(SECTION_NAME, jsonPayload);
  return Buffer.concat([base, newSection]);
}

/**
 * Read cleat.metadata from a WASM binary.
 */
function readMetadata(wasmBytes, filePath = "") {
  const found = findCustomSection(wasmBytes, SECTION_NAME, filePath);
  if (!found) return null;
  return JSON.parse(found.payload.toString("utf8"));
}

function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args.includes("--help") || args.includes("-h")) {
    console.log(`
Usage: node scripts/inject-metadata.js <wasm-file> [options]

Options:
  --name <name>         Workflow name (or CLEAT_WORKFLOW_NAME env var)
  --version <n>         Workflow version (or CLEAT_WORKFLOW_VERSION env var)
  --min-version <n>     Min compatible version (or CLEAT_MIN_COMPATIBLE_VERSION env var)
  --abi-version <n>     ABI version (or CLEAT_ABI_VERSION env var)
  --plugin-deps <json>  Plugin dependencies JSON (or CLEAT_PLUGIN_DEPS env var)
  --output, -o <file>   Output WASM path (default: overwrite input)
  --read                Read and display metadata instead of writing
  --verbose, -v         Verbose output
`);
    process.exit(0);
  }

  const readOnly = args.includes("--read");
  let wasmFile = null;
  const options = {};

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "--read" || arg === "--verbose" || arg === "-v") continue;
    if (arg === "--output" || arg === "-o") {
      options.output = args[++i];
    } else if (arg === "--name") {
      options.name = args[++i];
    } else if (arg === "--version") {
      options.version = parseInt(args[++i], 10);
    } else if (arg === "--min-version") {
      options.minVersion = parseInt(args[++i], 10);
    } else if (arg === "--abi-version") {
      options.abiVersion = parseInt(args[++i], 10);
    } else if (arg === "--plugin-deps") {
      options.pluginDeps = args[++i];
    } else if (!arg.startsWith("--")) {
      wasmFile = arg;
    }
  }

  if (!wasmFile) {
    console.error("Error: <wasm-file> is required");
    process.exit(1);
  }

  const wasmBytes = fs.readFileSync(wasmFile);

  if (readOnly) {
    const meta = readMetadata(wasmBytes, wasmFile);
    if (!meta) {
      console.error(`No '${SECTION_NAME}' section found in ${wasmFile}`);
      process.exit(1);
    }
    console.log(JSON.stringify(meta, null, 2));
    return;
  }

  // Build metadata from options or environment variables.
  const name =
    options.name ||
    process.env.CLEAT_WORKFLOW_NAME ||
    "unknown";
  const version =
    options.version !== undefined
      ? options.version
      : process.env.CLEAT_WORKFLOW_VERSION !== undefined
        ? parseInt(process.env.CLEAT_WORKFLOW_VERSION, 10)
        : 0;
  const minVersion =
    options.minVersion !== undefined
      ? options.minVersion
      : process.env.CLEAT_MIN_COMPATIBLE_VERSION !== undefined
        ? parseInt(process.env.CLEAT_MIN_COMPATIBLE_VERSION, 10)
        : 1;
  const abiVersion =
    options.abiVersion !== undefined
      ? options.abiVersion
      : process.env.CLEAT_ABI_VERSION !== undefined
        ? parseInt(process.env.CLEAT_ABI_VERSION, 10)
        : 1;
  const pluginDepsStr =
    options.pluginDeps || process.env.CLEAT_PLUGIN_DEPS || "{}";

  let pluginDeps = {};
  try {
    pluginDeps = JSON.parse(pluginDepsStr);
    if (typeof pluginDeps !== "object" || Array.isArray(pluginDeps)) {
      pluginDeps = {};
    }
  } catch {
    pluginDeps = {};
  }

  const meta = {
    workflow_name: name,
    workflow_version: version,
    min_compatible_version: minVersion,
    abi_version: abiVersion,
    plugin_deps: pluginDeps,
    sdk_language: "assemblyscript",
    sdk_version: "0.1.0",
    created_at: new Date().toISOString(),
  };

  const modified = injectMetadata(wasmBytes, meta, wasmFile);
  const outputPath = options.output || wasmFile;
  fs.writeFileSync(outputPath, modified);

  const isVerbose = args.includes("--verbose") || args.includes("-v");
  if (isVerbose) {
    console.log(`Stamped metadata into ${outputPath}`);
    console.log(`  workflow_name:           ${meta.workflow_name}`);
    console.log(`  workflow_version:        ${meta.workflow_version}`);
    console.log(`  min_compatible_version:  ${meta.min_compatible_version}`);
    console.log(`  abi_version:             ${meta.abi_version}`);
    if (Object.keys(meta.plugin_deps).length > 0) {
      console.log(`  plugin_deps:             ${JSON.stringify(meta.plugin_deps)}`);
    }
    const origSize = wasmBytes.length;
    const newSize = modified.length;
    console.log(`  Size: ${origSize} bytes -> ${newSize} bytes (+${newSize - origSize})`);
  } else {
    console.log(`Stamped cleat metadata into ${outputPath}`);
  }
}

if (require.main === module) {
  main();
}

module.exports = { injectMetadata, readMetadata, findCustomSection, buildCustomSection };

File: crates/cleat-sdk/src/bin/inject_metadata.rs
//! inject_metadata — Post-compile metadata stamping for Rust WASM workflows.
//!
//! Reads a compiled WASM binary (produced by `cargo build --target wasm32-wasip1`)
//! and injects a "cleat.metadata" custom section containing workflow identity
//! information. The host can read this metadata without instantiating the module.
//!
//! Usage:
//!   cargo run --bin inject_metadata -- <wasm-file> \
//!       --name PlaceOrder \
//!       --version 3 \
//!       [--output <output.wasm>]
//!
//! Environment variables are also supported:
//!   CLEAT_WORKFLOW_NAME, CLEAT_WORKFLOW_VERSION, CLEAT_MIN_COMPATIBLE_VERSION,
//!   CLEAT_ABI_VERSION, CLEAT_PLUGIN_DEPS

use std::env;
use std::fs;

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 || args.contains(&"--help".to_string()) || args.contains(&"-h".to_string()) {
        eprintln!(
            r#"Usage: inject_metadata <wasm-file> [options]

Options:
  --name <name>                 Workflow name (or CLEAT_WORKFLOW_NAME env var)
  --version <n>                 Workflow version (or CLEAT_WORKFLOW_VERSION env var)
  --min-version <n>             Min compatible version (or CLEAT_MIN_COMPATIBLE_VERSION env var)
  --abi-version <n>             ABI version (or CLEAT_ABI_VERSION env var)
  --plugin-deps <json>          Plugin dependencies JSON (or CLEAT_PLUGIN_DEPS env var)
  --child-binding-policy <str>  Child binding policy (or CLEAT_CHILD_BINDING_POLICY env var)
  --output, -o <file>           Output WASM path (default: overwrite input)
  --read                        Read and display metadata instead of writing
"#
        );
        std::process::exit(0);
    }

    let read_only = args.contains(&"--read".to_string());

    let mut wasm_file = None;
    let mut output_file = None;
    let mut name = None;
    let mut version = None;
    let mut min_version = None;
    let mut abi_version = None;
    let mut plugin_deps = None;
    let mut child_binding_policy: Option<String> = None;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--output" | "-o" => {
                i += 1;
                output_file = Some(args[i].clone());
            }
            "--name" => {
                i += 1;
                name = Some(args[i].clone());
            }
            "--version" => {
                i += 1;
                version = Some(args[i].parse::<u32>().expect("version must be a number"));
            }
            "--min-version" => {
                i += 1;
                min_version = Some(args[i].parse::<u32>().expect("min-version must be a number"));
            }
            "--abi-version" => {
                i += 1;
                abi_version = Some(args[i].parse::<u32>().expect("abi-version must be a number"));
            }
            "--plugin-deps" => {
                i += 1;
                plugin_deps = Some(args[i].clone());
            }
            "--child-binding-policy" => {
                i += 1;
                child_binding_policy = Some(args[i].clone());
            }
            "--read" => {}
            s if !s.starts_with("--") => {
                wasm_file = Some(args[i].clone());
            }
            _ => {
                eprintln!("Unknown option: {}", args[i]);
                std::process::exit(1);
            }
        }
        i += 1;
    }

    let wasm_file = wasm_file.expect("<wasm-file> is required");
    let wasm_bytes = match fs::read(&wasm_file) {
        Ok(bytes) => bytes,
        Err(e) => {
            eprintln!("Failed to read WASM file {}: {}", wasm_file, e);
            std::process::exit(1);
        }
    };

    if read_only {
        match read_metadata(&wasm_bytes) {
            Some(meta) => {
                println!("{}", serde_json::to_string_pretty(&meta).unwrap());
            }
            None => {
                eprintln!("No 'cleat.metadata' section found in {}", wasm_file);
                std::process::exit(1);
            }
        }
        return;
    }

    // Build metadata from args or environment variables.
    let resolved_name = name
        .or_else(|| env::var("CLEAT_WORKFLOW_NAME").ok())
        .unwrap_or_else(|| "unknown".to_string());

    let resolved_version = version
        .or_else(|| {
            env::var("CLEAT_WORKFLOW_VERSION")
                .ok()
                .map(|v| v.parse::<u32>().unwrap_or(0))
        })
        .unwrap_or(0);

    let resolved_min_version = min_version
        .or_else(|| {
            env::var("CLEAT_MIN_COMPATIBLE_VERSION")
                .ok()
                .map(|v| v.parse::<u32>().unwrap_or(1))
        })
        .unwrap_or(1);

    let resolved_abi_version = abi_version
        .or_else(|| {
            env::var("CLEAT_ABI_VERSION")
                .ok()
                .map(|v| v.parse::<u32>().unwrap_or(1))
        })
        .unwrap_or(1);

    let deps_str = plugin_deps
        .or_else(|| env::var("CLEAT_PLUGIN_DEPS").ok())
        .unwrap_or_else(|| "{}".to_string());

    let deps_map: std::collections::HashMap<String, String> =
        serde_json::from_str(&deps_str).unwrap_or_default();

    let resolved_child_binding_policy = child_binding_policy
        .or_else(|| env::var("CLEAT_CHILD_BINDING_POLICY").ok())
        .unwrap_or_else(|| "".to_string());

    let meta = serde_json::json!({
        "workflow_name": resolved_name,
        "workflow_version": resolved_version,
        "min_compatible_version": resolved_min_version,
        "abi_version": resolved_abi_version,
        "plugin_deps": deps_map,
        "child_binding_policy": resolved_child_binding_policy,
        "sdk_language": "rust",
        "language": "rust",
        "sdk_version": "0.1.0",
        "created_at": chrono_now(),
    });

    let modified = inject_metadata(&wasm_bytes, &meta);
    let output_path = output_file.unwrap_or(wasm_file);
    match fs::write(&output_path, modified) {
        Ok(()) => {}
        Err(e) => {
            eprintln!("Failed to write output to {}: {}", output_path, e);
            std::process::exit(1);
        }
    }

    eprintln!("Stamped cleat metadata into {}", output_path);
}

fn chrono_now() -> String {
    // Use a simple UTC timestamp without depending on the `chrono` crate.
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default();
    let secs = now.as_secs();
    // Format as ISO 8601
    let days = secs / 86400;
    let time_secs = secs % 86400;
    let hours = time_secs / 3600;
    let minutes = (time_secs % 3600) / 60;
    let seconds = time_secs % 60;

    // Simple date calculation from Unix epoch (1970-01-01).
    let (year, month, day) = days_to_date(days as i64);
    format!(
        "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z",
        year, month, day, hours, minutes, seconds
    )
}

fn days_to_date(mut days: i64) -> (i64, i64, i64) {
    // Algorithm from http://howardhinnant.github.io/date_algorithms.html
    days += 719468;
    let era = if days >= 0 { days } else { days - 146096 } / 146097;
    let doe = days - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    (y, m, d)
}

// ---- WASM binary manipulation ----

const SECTION_NAME: &str = "cleat.metadata";
const WASM_MAGIC: [u8; 4] = [0x00, 0x61, 0x73, 0x6D]; // \0asm
const WASM_VERSION: [u8; 4] = [0x01, 0x00, 0x00, 0x00];

fn encode_uleb128(mut value: u64) -> Vec<u8> {
    let mut result = Vec::new();
    loop {
        let mut b = (value & 0x7F) as u8;
        value >>= 7;
        if value != 0 {
            b |= 0x80;
        }
        result.push(b);
        if value == 0 {
            break;
        }
    }
    result
}

fn decode_uleb128(data: &[u8], offset: usize) -> Option<(u64, usize)> {
    let mut result: u64 = 0;
    let mut shift: u32 = 0;
    let mut consumed = 0;

    for &b in data.iter().skip(offset) {
        consumed += 1;
        result |= ((b & 0x7F) as u64) << shift;
        if b & 0x80 == 0 {
            return Some((result, consumed));
        }
        shift += 7;
        if shift > 63 {
            return None;
        }
    }
    None
}

fn build_custom_section(name: &str, payload: &[u8]) -> Vec<u8> {
    let name_bytes = name.as_bytes();
    let name_len = encode_uleb128(name_bytes.len() as u64);
    let mut full_payload = Vec::new();
    full_payload.extend_from_slice(&name_len);
    full_payload.extend_from_slice(name_bytes);
    full_payload.extend_from_slice(payload);

    let payload_len = encode_uleb128(full_payload.len() as u64);
    let mut section = vec![0x00]; // section ID 0 (custom)
    section.extend_from_slice(&payload_len);
    section.extend_from_slice(&full_payload);
    section
}

fn find_custom_section(wasm_bytes: &[u8], name: &str) -> Option<(Vec<u8>, usize, usize)> {
    if wasm_bytes.len() < 8 {
        return None;
    }
    if wasm_bytes[..4] != WASM_MAGIC || wasm_bytes[4..8] != WASM_VERSION {
        return None;
    }

    let name_bytes = name.as_bytes();
    let mut offset = 8;

    while offset < wasm_bytes.len() {
        let section_start = offset;
        let section_id = wasm_bytes[offset];
        offset += 1;

        if offset >= wasm_bytes.len() {
            break;
        }

        let (payload_len, consumed) = decode_uleb128(wasm_bytes, offset)?;
        offset += consumed;
        let section_end = offset + payload_len as usize;

        if section_end > wasm_bytes.len() {
            return None;
        }

        if section_id == 0 {
            // Custom section — check name.
            if let Some((name_len, nc)) = decode_uleb128(wasm_bytes, offset) {
                let name_start = offset + nc;
                let name_end = name_start + name_len as usize;
                if name_end <= section_end && &wasm_bytes[name_start..name_end] == name_bytes {
                    return Some((
                        wasm_bytes[name_end..section_end].to_vec(),
                        section_start,
                        section_end,
                    ));
                }
            }
        }

        offset += payload_len as usize;
    }

    None
}

fn inject_metadata(wasm_bytes: &[u8], meta: &serde_json::Value) -> Vec<u8> {
    let json_payload = serde_json::to_string(meta).unwrap();
    let json_bytes = json_payload.as_bytes();

    // Remove existing section if present.
    let base = match find_custom_section(wasm_bytes, SECTION_NAME) {
        Some((_, start, end)) => {
            let mut v = Vec::with_capacity(wasm_bytes.len() - (end - start));
            v.extend_from_slice(&wasm_bytes[..start]);
            v.extend_from_slice(&wasm_bytes[end..]);
            v
        }
        None => wasm_bytes.to_vec(),
    };

    let new_section = build_custom_section(SECTION_NAME, json_bytes);
    let mut result = base;
    result.extend_from_slice(&new_section);
    result
}

fn read_metadata(wasm_bytes: &[u8]) -> Option<serde_json::Value> {
    let (payload, _, _) = find_custom_section(wasm_bytes, SECTION_NAME)?;
    serde_json::from_slice(&payload).ok()
}
